File: 백준/Silver/1926. 그림/그림.py
import sys
input=sys.stdin.readline
from collections import deque

N,M = map(int, input().split())


painting = [list(map(int, input().split())) for _ in range(N)]
visited=[[0]*M for _ in range(N)]

queue = deque()
count=0

dx=[-1,1,0,0]
dy=[0,0,-1,1]

def bfs(i,j):
    area=1
    visited[i][j]=1
    queue.append((i,j))

    while queue:
        x,y = queue.popleft()
        for k in range(4):
            nx = x + dx[k]
            ny = y + dy[k]
            
            if 0<=nx<N and 0<=ny< M and visited[nx][ny]==0:
                if painting[nx][ny]==1:
                    visited[nx][ny]=1
                    area+=1
                    queue.append((nx,ny))
    return area
    
maxArea=0

for i in range(N):
    for j in range(M):
        if painting[i][j]==1 and visited[i][j]==0:
            count+=1
            area = bfs(i,j)
            maxArea = max(maxArea,area)

if count==0 :
    print(0)
    print(0)
else :
    print(count)
    print(maxArea)

File: 백준/Silver/1926. 그림/그림.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    public static class Node {
        private int x;
        private int y;

        public Node(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    public static void main(String[] args) throws IOException {


        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());


        int[] dx = {1, 0, -1, 0};
        int[] dy = {0, 1, 0, -1};
        int[][] board = new int[N][M];
        boolean[][] visited = new boolean[N][M];


        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        int count = 0;
        int maxA = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                Queue<Node> queue = new LinkedList<>();

                if (board[i][j] == 0 || visited[i][j])
                    continue;

                queue.offer(new Node(i, j));
                visited[i][j] = true;
                int area = 0;
                while (!queue.isEmpty()) {
                    area++;
                    Node node = queue.poll();
                    for (int d = 0; d < 4; d++) {
                        int x = node.x + dx[d];
                        int y = node.y + dy[d];

                        if (x >= 0 && x < N && y >= 0 && y < M) {
                            if (!visited[x][y] && board[x][y] == 1) {
                                queue.offer(new Node(x, y));
                                visited[x][y] = true;
                            }
                        }
                    }
                }
                maxA = Math.max(maxA,area);
                count++;
            }
        }

        System.out.println(count);
        System.out.println(maxA);
    }
}

